pydicom/filereader.py

```diff
@@ -1206,3 +1224,534 @@ def read_deferred_data_element(
+    following the BOT item (the first byte of the first Frame item).
+
+    Raises
+    ------
+    IOError
+        When file pointer is not positioned at first byte of first Frame item
+        after Basic Offset Table item or when parsing of Frame item headers
+        fails
+    ValueError
+        When the number of offsets doesn't match the specified number of frames
+
+    """
+    initial_position = fp.tell()
+    offset_values = []
+    current_offset = 0
+    i = 0
+    while True:
+        frame_position = fp.tell()
+        tag = TupleTag(fp.read_tag())
+        if int(tag) == SequenceDelimiterTag:
+            break
+        if int(tag) != ItemTag:
+            fp.seek(initial_position, 0)
+            raise IOError(
+                'Building Basic Offset Table (BOT) failed. Expected tag of '
+                f'Frame item #{i} at position {frame_position}.'
+            )
+        length = fp.read_UL()
+        if length % 2:
+            fp.seek(initial_position, 0)
+            raise IOError(
+                'Building Basic Offset Table (BOT) failed. '
+                f'Length of Frame item #{i} is not a multiple of 2.'
+            )
+        elif length == 0:
+            fp.seek(initial_position, 0)
+            raise IOError(
+                'Building Basic Offset Table (BOT) failed. '
+                f'Length of Frame item #{i} is zero.'
+            )
+
+        first_two_bytes = fp.read(2, 1)
+        if not fp.is_little_endian:
+            first_two_bytes = first_two_bytes[::-1]
+
+        # In case of fragmentation, we only want to get the offsets to the
+        # first fragment of a given frame. We can identify those based on the
+        # JPEG and JPEG 2000 markers that should be found at the beginning and
+        # end of the compressed byte stream.
+        if first_two_bytes in _START_MARKERS:
+            current_offset = frame_position - initial_position
+            offset_values.append(current_offset)
+
+        i += 1
+        fp.seek(length - 2, 1)  # minus the first two bytes
+
+    if len(offset_values) != number_of_frames:
+        raise ValueError(
+            'Number of frame items does not match specified Number of Frames.'
+        )
+    else:
+        basic_offset_table = offset_values
+
+    fp.seek(initial_position, 0)
+    return basic_offset_table
+
+
+class ImageFileReader(object):
+
+    """Reader for DICOM datasets representing Image Information Entities.
+
+    It provides efficient access to individual Frame items contained in the
+    Pixel Data element without loading the entire element into memory.
+
+    Attributes
+    ----------
+    filename: str
+        Path to the DICOM Part10 file on disk
+
+    Examples
+    --------
+    >>> from highdicom.io import ImageFileReader
+    >>> with ImageFileReader('/path/to/file.dcm') as image:
+    ...     print(image.metadata)
+    ...     for i in range(image.number_of_frames):
```

Comment:
I'm not a fan of the term `metadata` here, especially because it sounds too much like the File Meta Information defined for part 10 files in the Standard.  But also because it is really a Dataset (from `dcmread` on `open` call) of all data elements except the pixel data; pydicom (and DICOM itself) don't give unique status to pixel data, unlike other file types with a 'header' or 'metadata'; in DICOM they are all just data elements.

Having said that, I'm not sure what to call it.  In pydicom, using `stop_before_pixels=True` just gives a Dataset, simply with the pixel data element missing.  I'll mull this over a little more...

Reply:
In DICOMweb, the term [metadata](http://dicom.nema.org/medical/dicom/current/output/chtml/part18/sect_10.4.html#sect_10.4.1.1.2) is used to refer to the subset of Data Elements in a Data Set that are not considered [bulkdata](http://dicom.nema.org/medical/dicom/current/output/chtml/part18/sect_10.4.html#sect_10.4.1.1.5) (i.e., that have a size smaller than a defined threshold).

I generally use to term to indicate that a given `pydicom.dataset.Dataset` may not contain the Pixel Data element.

Reply:
I have heard that term too, and have also been confused by it. I think it comes from image formats like TIFF who have embedded metadata tags.
What I encounter more often is "DICOM header data" - not sure if that would be a better term.

Reply:
> DICOM header data

I think 'header' applies to other formats where the pixel data is not tagged, but whatever is left after some initial data is just raw pixel bytes.  It assumes every file is basically pixel data, plus a few extra things.

In DICOM, many file types (SOP Classes) do not have pixel data, so the whole file would be 'header', which doesn't really make sense.  Same goes for `metadata` in the sense of everything that is not pixels.

> could we have something like a `frames` property as an alternate to `pixel_array`?

Any thoughts on the above general comment?  If this could work as an add-on to `Dataset`, then the nomenclature issue could go away.

Reply:
> In DICOM, many file types (SOP Classes) do not have pixel data, so the whole file would be 'header', which doesn't really make sense.

Fair enough. I didn't really like the term too much, was just thinking of an alternative to metadata.

> Any thoughts on the above general comment? 

I think it makes sense. Frames are an inherent property of DICOM files, so it would fit IMHO. Also, I think that the ability to read single frames is an important feature that belongs into the core.

Reply:
Regarding terminology, I think that anyone who uses DICOMweb will be familiar with the term "metadata" and the value returned by the `metadata` property will (more or less) correspond to what one would get via a DICOMweb `/metadata` call. Similarly, the value returned by `read_frame()` corresponds to what one would get via a DICOMweb `/frames/{frameNumber}` call.